flash_io.c

```diff
@@ -0,0 +1,622 @@
+} flash_data_t;
+
+// flash_data_t must be the same size as a page. It is required by the algorithm
+static_assert(sizeof(flash_data_t) == FLASH_PAGE_SIZE, "FLASH ERROR: sizeof(flash_data_t) != FLASH_PAGE_SIZE\n");
+
+typedef struct
+{
+  const flash_data_t* ptr;
+  int8_t              current_index;
+} flash_ptr_t;
+
+static const init_targets
+target = {
+
+#if CONFIG_MAGNETRON
+  .magnetron = 1,
+#endif
+
+#if CONFIG_WATER
+  .water = 1,
+#endif
+
+#if CONFIG_AUTO == CONFIG_AUTO_PILOT || CONFIG_AUTO == CONFIG_AUTO_MAPPER
+  .pilot = 1,
+#endif
+
+#if CONFIG_AUTO == CONFIG_AUTO_MAPPER
+  .mapper = 1,
+#endif
+
+#if CONFIG_SHUTTER
+  .shutter = 1,
+#endif
+
+#if CONFIG_STIRER
+  .stirer = 1,
+#endif
+
+};
+
+static flash_ptr_t flash_ptr_1 =
+  {
+    .ptr           = (flash_data_t*)&FLASH_CONFIG_DATA_START,
+    .current_index = -1,
+  };
+
+static flash_ptr_t flash_ptr_2 =
+  {
+    .ptr           = (void*)&FLASH_CONFIG_DATA_START + FLASH_SECTOR_SIZE,
+    .current_index = -1,
+  };
+
+static flash_valid_data_t flash_last_written =
+  {
+    .targets = target,
+    .tag     = TAG,
+  };
+
+static flash_data_t flash_current =
+  {
+    .data.targets = target,
+    .data.tag     = TAG,
+  };
+
+static bool
+flash_is_page_empty(const int8_t* ptr)
+{
+  for(int i = 0; i < FLASH_PAGE_SIZE; i++)
+  {
+    if(*ptr != -1)
+      return false;
+    ptr++;
+  }
+  return true;
+}
+
+static void
+flash_find_last_blocks()
+{
+  bool empty_1 = true;
+  bool empty_2 = true;
+
+  for(int i = FLASH_MAX_PAGE_INDEX; i >= 0; i--)
+  {
+    const int8_t* ptr_1 = (int8_t*)&flash_ptr_1.ptr[i];
+    const int8_t* ptr_2 = (int8_t*)&flash_ptr_2.ptr[i];
+
+    if(empty_1)
+    {
+      if(!flash_is_page_empty(ptr_1))
+      {
+        flash_ptr_1.current_index = i;
+        empty_1 = false;
+      }
+    }
+
+    if(empty_2)
+    {
+      if(!flash_is_page_empty(ptr_2))
+      {
+        flash_ptr_2.current_index = i;
+        empty_2 = false;
+      }
+    }
+  }
+}
+
+static void
+flash_read(flash_ptr_t* flash_ptr_, furnace_context_t* ctx)
+{
+  const flash_valid_data_t* flash_ptr = &flash_ptr_->ptr[flash_ptr_->current_index].data;
+
+  if(flash_ptr->targets.as_unsigned != target.as_unsigned)
+    return;
+
+  ctx->log_bits    = flash_ptr->log_bits;
+  ctx->pwm_level   = flash_ptr->pwm_level;
+  ctx->ceiling_pwm = flash_ptr->ceiling_pwm;
+
+#if CONFIG_WATER
+  ctx->pwm_water = flash_ptr->pwm_water;
+#endif
+
+#if CONFIG_MAGNETRON
+  ctx->pulse_count        = flash_ptr->magnetron_pulse_count;
+  ctx->magnetron_deadline = flash_ptr->magnetron_magnetron_deadline;
+#endif
+
+#if CONFIG_AUTO == CONFIG_AUTO_PILOT || CONFIG_AUTO == CONFIG_AUTO_MAPPER
+  ctx->pilot.des_temp   = flash_ptr->pilot_des_temp;
+  ctx->pilot.is_enabled = flash_ptr->pilot_is_enabled;
+#endif
+
+#if CONFIG_AUTO == CONFIG_AUTO_MAPPER
+  ctx->mapper.max_pwm_temp = flash_ptr->mapper_max_pwm_temp;
+  ctx->mapper.is_enabled   = flash_ptr->mapper_is_enabled;
+#endif
```

Comment:
What happens if we saved the data with one config but later try to read it with a different one?
For example, we saved to the flash when `CONFIG_MAGNETRON = 1`, but later programmed new binary with `CONFIG_MAGNETRON = 0`.

Perhaps `flash_data_t` has to have a field saying which fields were serialized and reject loading if configs mismatch or should try to fallback to common subset, but the latter might be too complicated and not really used frequently.

Config mismatch may happen if we reuse the same pico board for different targets, therefore we should at least detect that condition and reject loading the old state from flash.

Additionally, I think we should probably clear flash when programming new image. We can give a option to make/cmake to skip the erasure, but by default, to make sure we don't start reading bogus values, we should just erase the old state on the new image.

Reply:
The option to force a flash reset is a good idea, and I'd like to implement it in a separate PR. Alternatively, I could make it the default action when a config mismatch occurs. However, I don’t think this should happen by default—it’s up to you to decide how we want this to work.

In `init_flash()` (more specifically, in `flash_validate_blocks()` ), we check for configuration mismatches. If there's a mismatch, we simply skip reading from memory. The `targets` field is responsible for storing information about the config.

Reply:
I think we should reset the flash if targets between what binary has configured and what is stored in the flash differ. Otherwise we would read corrupted values.